### PR TITLE
VC-4250: added MsgPayments

### PR DESCRIPTION
**Pull request by at-kh**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		return err
+	}
+
+	var fee uint64
+
+	switch isChargeFee(accountFrom[0], accountTo[0], walletFrom[0].Kind, walletTo[0].Kind) {
+	case false:
+		fee = 0
+	default:
+		asset := m.getAsset(ctx, msg.Asset)
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go]: asset можно запросить из нашей БД, не за чем блокчейн дергать.

	asset, err := m.assetRepo.GetAssets(filter.NewFilter().SetArgument(dbtypes.FieldName, msg.Asset))
	switch {
	case err != nil:
		return err
	case len(asset) != 1:
		return types.ErrInvalidAddressField
	}

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		if asset.State != assets.ASSET_STATE_ACTIVE {
+			return banking.ErrBadAssetIsNotActive
+		}
+
+		// check asset policy
+		if !asset.CheckPolicy(assets.ASSET_POLICY_TRANSFERABLE) {
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go]: Если нам сообщение пришло, то оно на блокчейне прошло без ошибок, можно убрать проверку

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	juno "github.com/forbole/juno/v2/types"
+
+	dbtypes "github.com/forbole/bdjuno/v2/database/types"
+)
+
+// handleMsgPayments allows to properly handle a MsgSetState
+func (m *Module) handleMsgPayments(goCtx context.Context, tx *juno.Tx, index int, msg *banking.MsgPayment) error {
+	ctx := sdk.UnwrapSDKContext(goCtx)
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go]: ctx не нужен, так как asset из нашей БД доставаться будет.

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	if err := m.bankingRepo.SaveMsgPayments(msg); err != nil {
+		return err
+	}
+
+	payments, err := m.bankingRepo.GetPayments(filter.NewFilter().SetArgument(dbtypes.FieldCreator, msg.Creator))
+	if err != nil {
+		return err
+	}
+
+	if len(payments) != 1 {
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go]: Можно заменить вот такой проверкй:

	switch {
	case err != nil:
		return err
	case len(payments) != 1:
		return types.ErrInvalidCreatorField
	}

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+		return banking.ErrInvalidCreatorField
+	}
+
+	time, err := time.Parse(time.RFC3339, tx.Timestamp)
+	if err != nil {
+		return err
+	}
+
+	accountFrom, err := m.accountsRepo.GetAccounts(filter.NewFilter().SetArgument(dbtypes.FieldAddress, msg.WalletFrom))
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go:22]: Добавь проверку на то что масив не пустой

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return err
+	}
+
+	accountTo, err := m.accountsRepo.GetAccounts(filter.NewFilter().SetArgument(dbtypes.FieldAddress, msg.WalletTo))
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go:30]: Добавь проверку на то что масив не пустой

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return err
+	}
+
+	walletFrom, err := m.walletsRepo.GetWallets(filter.NewFilter().SetArgument(dbtypes.FieldAddress, msg.WalletFrom))
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go:38]: Добавь проверку на то что масив не пустой

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return err
+	}
+
+	walletTo, err := m.walletsRepo.GetWallets(filter.NewFilter().SetArgument(dbtypes.FieldAddress, msg.WalletTo))
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go:46]: Добавь проверку на то что масив не пустой

File: modules/vipcoin/chain/banking/msg_payment.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	var fee uint64
+
+	switch isChargeFee(accountFrom[0], accountTo[0], walletFrom[0].Kind, walletTo[0].Kind) {
+	case false:
+		fee = 0
+	default:
+		asset := m.getAsset(ctx, msg.Asset)
+		if asset.State != assets.ASSET_STATE_ACTIVE {
```
Comment (alex-bodnar) [inline modules/vipcoin/chain/banking/msg_payment.go]: Тоже проверку можно удалить

**Review by alex-bodnar — APPROVED** [review]
LGTM

**Review by IamStubborN — APPROVED** [review]
LGTM

**Pull request merged**